### ISSUE TITLE
bug: Windows service deadlocks on thread exit

### ISSUE THREAD

**Issue by muffins** [facebook/osquery#4235]
An issue we've experienced internally is that if the osquery service is restarted very quickly, say with `Restart-Service osqueryd`, and one has the Windows Event Log publisher turned on, there seem to be residual osquery processes that get left behind and prevent a new instance of the service from coming back online. I'm currently triaging, but folks can work around this by _not_ using `Restart-Service`, and instead take the slower route of
```
Stop-Service osqueryd
Get-Process | Where-Object {$_.ProcessName -eq 'osqueryd'} | Stop-Process -Force
Start-Sleep -s 1
Start-Service osqueryd
```

To reproduce locally, one can leverage the following steps:
```
1.) install osquery as a service on your box

2.) copy and paste the following:

  C:\Users\Nicholas\work\repos\osquery [win-zombie-service-proc ≡]
  λ  cp .\tools\tests\test_client* C:\ProgramData\osquery\
  C:\Users\Nicholas\work\repos\osquery [win-zombie-service-proc ≡]
  λ  cp .\tools\tests\test_server_ca.pem C:\ProgramData\osquery\
  C:\Users\Nicholas\work\repos\osquery [win-zombie-service-proc ≡]
  λ  cp .\tools\tests\test_enroll_secret.txt C:\ProgramData\osquery\

And set your `osquery.flags` to be this:

  --allow_unsafe=true
  --config_plugin=tls
  --config_tls_endpoint=/config
  --config_tls_refresh=180
  --enable_monitor
  --enable_mouse_events
  --enroll_secret_path=C:\ProgramData\osquery\test_enroll_secret.txt
  --enroll_tls_endpoint=/enroll
  --tls_hostname=localhost
  --tls_server_certs=C:\ProgramData\osquery\test_server_ca.pem
  --distributed_plugin=tls
  --distributed_tls_read_endpoint=/distributed_read
  --distributed_tls_write_endpoint=/distributed_write
  --disable_extensions=true
  --extensions_interval=3
  --extensions_timeout=3
  --extensions_autoload=C:\ProgramData\osquery\extensions.load
  --logger_plugin=tls
  --logger_tls_compress
  --logger_tls_endpoint=/log
  --logger_tls_period=300
  --logger_tls_max=4194304
  --pidfile=C:\ProgramData\osquery\osquery.pid
  --database_path=C:\ProgramData\osquery\osquery.db
  --disable_events=false
  --events_expiry=300
  --windows_event_channels=System,Application,Setup,Security,Microsoft-Windows-PowerShell/Operational

You can make use of `--verbose` if you feel the need, but this might not be super helpful in debugging the service.

3.) In a separate shell, start up the Test HTTP server:

  C:\Users\Nicholas\work\repos\osquery [win-zombie-service-proc ≡]
  λ  python .\tools\tests\test_http_server.py --persist --verbose --tls 443
  -- [DEBUG] Starting TLS/HTTPS server on TCP port: 443


4.) Verify that osqueryd is able to talk with your test server in a different shell:

  C:\Users\Nicholas
  λ  C:\ProgramData\osquery\osqueryd\osqueryd.exe --flagfile=C:\ProgramData\osquery\osquery.flags
  I0406 15:26:47.441854   508 init.cpp:388] osquery initialized [version=3.2.2-4-g3abadc77]
  I0406 15:26:47.477951   508 system.cpp:344] Found stale process for osqueryd (1904)
  I0406 15:26:47.477951   508 system.cpp:377] Writing osqueryd pid (10860) to C:\ProgramData\osquery\osquery.pid
  I0406 15:26:47.775974 12840 watcher.cpp:549] osqueryd watcher (10860) executing worker (12476)
  I0406 15:26:47.798035  3636 init.cpp:385] osquery worker initialized [watcher=10860]
  I0406 15:26:47.800040  3636 rocksdb.cpp:132] Opening RocksDB handle: C:\ProgramData\osquery\osquery.db
  I0406 15:26:47.870000  3636 init.cpp:624] Cannot start extension manager: Extensions disabled
  I0406 15:26:47.871017  3636 tls.cpp:241] TLS/HTTPS POST request to URI: https://localhost/config
  I0406 15:26:49.249413  3636 http_client.cpp:51] SSL SHORT_READ_ERROR: http_client closing socket
  I0406 15:26:50.251448  3636 tls_enroll.cpp:59] TLSEnrollPlugin requesting a node enroll key from: https://localhost/enroll
  I0406 15:26:51.288216  3636 tls.cpp:241] TLS/HTTPS POST request to URI: https://localhost/enroll
  I0406 15:26:52.298629  3636 http_client.cpp:51] SSL SHORT_READ_ERROR: http_client closing socket
  I0406 15:26:52.300634  3636 tls.cpp:241] TLS/HTTPS POST request to URI: https://localhost/config
  I0406 15:26:53.309578  3636 http_client.cpp:51] SSL SHORT_READ_ERROR: http_client closing socket
  W0406 15:26:53.336649  3636 windows_event_log.cpp:53] Failed to subscribe to Security: 5
  I0406 15:26:53.337651 13220 events.cpp:746] Starting event publisher run loop: windows_events
  I0406 15:26:53.337651  3636 main.cpp:107] Not starting the distributed query service: Distributed query service not enabled.
  I0406 15:26:53.338655  3252 scheduler.cpp:75] Executing scheduled query tls_proc: select * from processes
  I0406 15:26:53.446965  3252 virtual_table.cpp:300] Error casting disk_bytes_read () to BIGINT
  I0406 15:26:53.446965  3252 virtual_table.cpp:300] Error casting disk_bytes_written () to BIGINT
  I0406 15:26:53.447944  3252 virtual_table.cpp:300] Error casting disk_bytes_read () to BIGINT
  I0406 15:26:53.447944  3252 virtual_table.cpp:300] Error casting disk_bytes_written () to BIGINT
  ...


5.) Now, start up the web server and monitor it as you start up the osqueryd service. Note that the service installation should be the default, the results of sc.exe qc are:

  C:\Users\Nicholas
  λ  sc.exe qc osqueryd
  [SC] QueryServiceConfig SUCCESS

  SERVICE_NAME: osqueryd
          TYPE               : 10  WIN32_OWN_PROCESS
          START_TYPE         : 2   AUTO_START
          ERROR_CONTROL      : 1   NORMAL
          BINARY_PATH_NAME   : C:\ProgramData\osquery\osqueryd\osqueryd.exe --flagfile=\ProgramData\osquery\osquery.flags
          LOAD_ORDER_GROUP   :
          TAG                : 0
          DISPLAY_NAME       : osqueryd
          DEPENDENCIES       :
          SERVICE_START_NAME : LocalSystem


6.) Once you see your node checking in with the server, you can use `Restart-Service osqueryd` to trigger the bug. It's worth noting, this didn't actually seem to connect to the server, however I was able to trigger the bug with this.
```

**Comment by muffins** [comment]
I believe the bigger issue here is why didn't the worker kill itself due to lack of watcher.

**Comment by muffins** [comment]
This deadlock has to do with numerous services being spun up, thus not necessarily related to _just_ events.

**Comment by ExceptionalHandler** [comment]
This issue seems to occurs because the APC scheduled in lib\boost-msvc14\local\include\boost\asio\detail\impl\win_thread.ipp (line -54) never gets scheduled. 

This could be because the APC was scheduled while destroying a thread_local http::Client object (created at osquery\remote\transports\tls.cpp line:170) while the thread is being terminated. 
This results in QueueuserAPC getting called from code being executed after a loader lock was taken by LdrShutdownThread. There is no documented guidance about not calling QueuUserAPC from within a thread's termination code, but having that code called from within a Loader lock that will be released after the APC is delivered is an uncomfortable scenario.  

Below is the stack of the thread just after calling QueuUserAPC. 
 
```
# Child-SP          RetAddr           Call Site
00 0000009b`3e7ff728 00007ffc`60833b7f ntdll!NtWaitForSingleObject+0x14
01 0000009b`3e7ff730 00007ff6`f780390c KERNELBASE!WaitForSingleObjectEx+0x9f
02 0000009b`3e7ff7d0 00007ff6`f78058ea osqueryd!boost::asio::detail::win_thread::join+0x7c [c:\programdata\chocolatey\lib\boost-msvc14\local\include\boost\asio\detail\impl\win_thread.ipp @ 57] 
03 0000009b`3e7ff810 00007ff6`f77ff2b6 osqueryd!boost::asio::detail::win_iocp_io_context::shutdown+0x22a [c:\programdata\chocolatey\lib\boost-msvc14\local\include\boost\asio\detail\impl\win_iocp_io_context.ipp @ 132] 
04 (Inline Function) --------`-------- osqueryd!boost::asio::detail::service_registry::shutdown_services+0x2a [c:\programdata\chocolatey\lib\boost-msvc14\local\include\boost\asio\detail\impl\service_registry.ipp @ 44] 
05 (Inline Function) --------`-------- osqueryd!boost::asio::execution_context::shutdown+0x2a [c:\programdata\chocolatey\lib\boost-msvc14\local\include\boost\asio\impl\execution_context.ipp @ 41] 
06 0000009b`3e7ff8c0 00007ff6`f77fef93 osqueryd!boost::asio::execution_context::~execution_context+0x46 [c:\programdata\chocolatey\lib\boost-msvc14\local\include\boost\asio\impl\execution_context.ipp @ 34] 
07 (Inline Function) --------`-------- osqueryd!boost::asio::io_context::{dtor}+0x20 [c:\programdata\chocolatey\lib\boost-msvc14\local\include\boost\asio\impl\io_context.ipp @ 57] 
08 0000009b`3e7ff900 00007ff6`f75cafa5 osqueryd!osquery::http::Client::~Client+0x103 [c:\git\facebook\osquery\osquery\remote\http_client.h @ 238] 
09 0000009b`3e7ff950 00007ff6`f7c425f7 osqueryd!std::_Ref_count_base::_Decref+0x35 [c:\program files (x86)\microsoft visual studio 14.0\vc\include\memory @ 113] 
0a 0000009b`3e7ff980 00007ffc`63631493 osqueryd!__dyn_tls_dtor+0x67 [f:\dd\vctools\crt\vcstartup\src\tls\tlsdtor.cpp @ 119] 
0b 0000009b`3e7ff9c0 00007ffc`636315ad ntdll!LdrpCallInitRoutine+0x6f
0c 0000009b`3e7ffa30 00007ffc`6362ed18 ntdll!LdrpCallTlsInitializers+0x8d
0d 0000009b`3e7ffab0 00007ffc`636677fe ntdll!LdrShutdownThread+0x1c8
0e 0000009b`3e7ffbb0 00007ffc`60891983 ntdll!RtlExitUserThread+0x3e
0f 0000009b`3e7ffbf0 00007ffc`60cd123a KERNELBASE!FreeLibraryAndExitThread+0x43
10 0000009b`3e7ffc20 00007ff6`f7cae81d KERNEL32!FreeLibraryAndExitThreadStub+0xa
11 0000009b`3e7ffc50 00007ff6`f7caeaf9 osqueryd!common_end_thread+0x69 [minkernel\crts\ucrt\src\appcrt\startup\thread.cpp @ 274] 
12 0000009b`3e7ffc80 00007ff6`f7cae677 osqueryd!_endthreadex+0x9 [minkernel\crts\ucrt\src\appcrt\startup\thread.cpp @ 289] 
13 0000009b`3e7ffcb0 00007ffc`60cd1fe4 osqueryd!thread_start<unsigned int (__cdecl*)(void * __ptr64)>+0x57 [minkernel\crts\ucrt\src\appcrt\startup\thread.cpp @ 115] 
14 0000009b`3e7ffce0 00007ffc`6368efc1 KERNEL32!BaseThreadInitThunk+0x14
15 0000009b`3e7ffd10 00000000`00000000 ntdll!RtlUserThreadStart+0x21
```

One solution is to call boost::asio::detail::win_thread::set_terminate_threads(true) if we decide to create thread_local instances of any classes that consist of boost::asio::io_context  object.

**Comment by muffins** [comment]
@ExceptionalHandler awesome investigation, thanks for posting your findings! I'll try that out and see if it helps, and that hopefully explains what's up with the TLS config and logger plugins aiding the deadlock, however we can still repro the deadlock with local FS configuration and logging, with events and extensions enabled, any thoughts on those? The stack traces for the threads in these circumstances are typically all threads blocking on _endthreadex or the like, and ultimately follow a similar path into the kernel before hitting KeSynchronizeExecution

**Comment by ExceptionalHandler** [comment]
@muffins Thanks for the ack. 
I could not repro the issue with distributed plugins and events enabled without TLS. A look at the thread stack / dump could confirm if that is a same issue or a different one. We can also use the boost::asio flag mentioned above to confirm if the two issues are identical.

**Comment by muffins** [comment]
This was fixed in #4286
